### PR TITLE
build!: drop node8 and convert to TypeScript

### PR DESCRIPTION
**Pull request by alexander-fenster**
This is the final library to be migrated to TypeScript generator! :)

Long story short:
- Node.js v8 support is dropped
- `gts` v2
- getting rid of `proto/` folder in favor of auto-generated `protos/` folder

One code changes: in the manually written enums string values were used:

```ts
  const enum TypeCode {	
    TYPE_CODE_UNSPECIFIED = 'TYPE_CODE_UNSPECIFIED',	
    BOOL = 'BOOL',	
    INT64 = 'INT64',	
   ...  
}
```

I'm getting rid of manually written types so this gets replaced with a real enum (with integer values), hence some smaller code changes in `src/codec.ts`. Luckily, `src/codec.ts` is covered by tests really well, so (a) this problem was caught, and (b) I'm pretty confident in a fix.

No changes to system tests or samples, which makes me feel really good :)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,18 +33,15 @@
     "samples-test": "cd samples/ && npm link ../ && npm test && cd ../",
     "system-test": "mocha build/system-test --timeout 600000",
     "cleanup": "mocha scripts/cleanup.js --timeout 30000",
-    "test": "c8 mocha build/test build/test/common",
+    "test": "mocha build/test build/test/common",
```
Comment (JustinBeckwith) [inline package.json:36]: What happened to `c8`?

File: package.json [hunk 2]
```diff
@@ -34,18 +33,15 @@
...
-    "compile": "tsc -p . && cp -r src/v1 build/src && cp -r protos build && cp -r proto build && cp test/*.js build/test",
+    "compile": "tsc -p . && cp -r protos build",
     "prepare": "npm run compile",
     "pretest": "npm run compile",
     "presystem-test": "npm run compile",
-    "proto": "npm run proto:spanner && npm run proto:spanner_instance_admin && npm run proto:spanner_database_admin",
-    "proto:spanner": "mkdir -p proto && pbjs -t static-module -w commonjs -p node_modules/google-proto-files google/spanner/v1/spanner.proto | pbts -o proto/spanner.d.ts -",
-    "proto:spanner_instance_admin": "pbjs -t static-module -w commonjs -p node_modules/google-proto-files google/spanner/admin/instance/v1/spanner_instance_admin.proto | pbts -o proto/spanner_instance_admin.d.ts -",
-    "proto:spanner_database_admin": "pbjs -t static-module -w commonjs -p node_modules/google-proto-files google/spanner/admin/database/v1/spanner_database_admin.proto | pbts -o proto/spanner_database_admin.d.ts -",
+    "proto": "compileProtos src",
```
Comment (JustinBeckwith) [inline package.json:44]: I love that

File: package.json [hunk 3]
```diff
@@ -62,7 +58,8 @@
     "checkpoint-stream": "^0.1.1",
     "events-intercept": "^2.0.0",
     "extend": "^3.0.2",
-    "google-gax": "^1.7.5",
+    "google-auth-library": "^6.0.0",
```
Comment (JustinBeckwith) [inline package.json:61]: I don't like that this is explicit, but we can clean up later

File: src/database.ts [hunk 4]
```diff
@@ -1991,7 +1993,7 @@ class Database extends GrpcServiceObject {
 
     const getWriteSession = this.pool_.getWriteSession.bind(this.pool_);
     // Loop to retry 'Session not found' errors.
-    while (true) {
+    for (;;) {
```
Comment (JustinBeckwith) [inline src/database.ts:1996]: That is far less obvious.  Can we put it back?

File: src/index.ts [hunk 5]
```diff
@@ -61,109 +56,11 @@ export interface SpannerOptions extends GrpcClientOptions {
 export interface RequestConfig {
   client: string;
   method: string;
-  // tslint:disable-next-line: no-any
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
   reqOpts: any;
   gaxOpts?: {};
 }
-/*!
```
Comment (JustinBeckwith) [inline src/index.ts:68]: uh ... won't that be bad for the generated docs?

File: tsconfig.json [hunk 6]
```diff
@@ -4,7 +4,11 @@
     "rootDir": ".",
     "outDir": "build",
     "noImplicitAny": false,
-    "resolveJsonModule": true
+    "resolveJsonModule": true,
+    "lib": [
+      "es2016",
```
Comment (JustinBeckwith) [inline tsconfig.json:9]: *2018

File: package.json (hunk 1 above)
Comment (alexander-fenster) [reply package.json:36]: Putting it back! It spammed my terminal window and got punished.

File: .gitignore [hunk 7]
```diff
@@ -10,7 +11,4 @@ system-test/*key.json
 *.lock
 .DS_Store
 package-lock.json
-.vscode
-yarn.lock
 __pycache__
-.idea
```
Comment (xiaozhenliu-gg5) [inline .gitignore:16]: can we add one line here? :)

File: .jsdoc.js [hunk 8]
```diff
@@ -31,18 +34,19 @@ module.exports = {
...
     includePattern: '\\.js$'
   },
   templates: {
-    copyright: 'Copyright 2019 Google, LLC.',
+    copyright: 'Copyright 2020 Google LLC',
     includeDate: false,
     sourceFiles: false,
-    systemName: '@google-cloud/spanner',
+    systemName: 'spanner',
     theme: 'lumen',
```
Comment (xiaozhenliu-gg5) [inline .jsdoc.js:47]: Just curious, we decided to rename it here to `spanner`?

File: package.json (hunk 3 above)
Comment (alexander-fenster) [reply package.json:61]: This was used directly somewhere and `eslint` yelled at me, so I added it here. Same with other dependencies added.

File: src/database.ts (hunk 4 above)
Comment (alexander-fenster) [reply src/database.ts:1996]: Nope - it's an `eslint` rule :) Do you want to make a `// eslint-ignore` exception here?

File: src/index.ts (hunk 5 above)
Comment (alexander-fenster) [reply src/index.ts:68]: I don't know! Should be fine... probably.

File: tsconfig.json (hunk 6 above)
Comment (alexander-fenster) [reply tsconfig.json:9]: Fixed!

File: .gitignore (hunk 7 above)
Comment (alexander-fenster) [reply .gitignore:16]: I added it :) The "no new line" thingy is red, not green! :)

File: .jsdoc.js (hunk 8 above)
Comment (alexander-fenster) [reply .jsdoc.js:47]: Hmmmmm

Comment (alexander-fenster) [reply .jsdoc.js:47]: Nice catch! It's the missing `package-name`. Fixed!

File: .gitignore (hunk 7 above)
Comment (xiaozhenliu-gg5) [reply .gitignore:16]: Thank you!

File: .jsdoc.js (hunk 8 above)
Comment (xiaozhenliu-gg5) [reply .jsdoc.js:47]: Thank you! congrats for the last API conversion! :)

File: src/database.ts (hunk 4 above)
Comment (JustinBeckwith) [reply src/database.ts:1996]: I think that would be better?  But I'm not blocking on it or anything :)

Comment (alexander-fenster) [reply src/database.ts]: Reverted!

File: package.json [hunk 9]
```diff
@@ -38,14 +37,11 @@
     "ycsb": "node ./benchmark/ycsb.js  run -P ./benchmark/workloada -p table=usertable -p cloudspanner.instance=ycsb-instance -p operationcount=100 -p cloudspanner.database=ycsb",
     "fix": "gts fix && eslint --fix '**/*.js'",
```
Comment (bcoe) [inline package.json:38]: we shouldn't need the `eslint` any longer with the new `gts`.

File: package.json [hunk 10]
```diff
@@ -95,25 +92,31 @@
...
     "concat-stream": "^2.0.0",
     "dedent": "^0.7.0",
+    "duplexify": "^4.1.1",
     "eslint": "^6.0.0",
     "eslint-config-prettier": "^6.0.0",
     "eslint-plugin-node": "^11.0.0",
     "eslint-plugin-prettier": "^3.0.0",
-    "gts": "^1.0.0",
+    "execa": "^4.0.0",
+    "gts": "2.0.0-alpha.9",
```
Comment (bcoe) [inline package.json:101]: @JustinBeckwith if you give me broad strokes of what's needed, I might be able to help draft the migration guide.

File: synth.py [hunk 11]
```diff
@@ -8,72 +8,48 @@
...
-gapic = gcp.GAPICGenerator()
-
-spanner = gapic.node_library(
+spanner = gapic.typescript_library(
     'spanner', 'v1',
-    config_path='/google/spanner/artman_spanner.yaml')
-
-spanner_admin_database = gapic.node_library(
+    proto_path='google/spanner/v1',
+    generator_args={
```
Comment (bcoe) [inline synth.py:16]: as Justin mentioned, awesome cleanup in here.

File: package.json (hunk 9 above)
Comment (alexander-fenster) [reply package.json:38]: Removing it!

**Pull request merged**